r/src/r_to_arrow.cpp

```diff
@@ -0,0 +1,814 @@
+  }
+};
+
+template <typename T>
+class RListConverter;
+
+template <typename U, typename Enable = void>
+class RDictionaryConverter;
+
+template <typename U>
+class RDictionaryConverter<U, enable_if_has_c_type<U>>
+    : public DictionaryConverter<U, RConverter> {
+ public:
+  Status Append(RScalar* value) override {
+    return Status::NotImplemented(
+        "dictionaries only implemented with string value types");
+  }
+};
+
+template <typename U>
+class RDictionaryConverter<U, enable_if_has_string_view<U>>
+    : public DictionaryConverter<U, RConverter> {
+ public:
+  Status Append(RScalar* value) override {
+    if (RValue::IsNull(value)) {
+      return this->value_builder_->AppendNull();
+    } else {
+      ARROW_RETURN_NOT_OK(
+          RValue::Convert(this->value_type_, this->options_, value, view_));
+      return this->value_builder_->Append(view_.bytes, static_cast<int32_t>(view_.size));
+    }
+  }
+
+ protected:
+  RBytesView view_;
+};
+
+template <typename T, typename Enable = void>
+struct RConverterTrait;
+
+template <typename T>
+struct RConverterTrait<
+    T, enable_if_t<!is_nested_type<T>::value && !is_interval_type<T>::value &&
+                   !is_extension_type<T>::value>> {
+  using type = RPrimitiveConverter<T>;
+};
+
+template <typename T>
+struct RConverterTrait<T, enable_if_list_like<T>> {
+  using type = RListConverter<T>;
+};
+
+template <typename T>
+class RListConverter : public ListConverter<T, RConverter, RConverterTrait> {
+ public:
+  Status Append(RScalar* value) {
+    if (RValue::IsNull(value)) {
+      return this->list_builder_->AppendNull();
+    }
+
+    // append one element to the list
+    RETURN_NOT_OK(this->list_builder_->Append());
+
+    // append the contents through the list value converter
+    SEXP obj = *reinterpret_cast<SEXP*>(value->data);
+    R_xlen_t size = XLENGTH(obj);
+    RETURN_NOT_OK(this->list_builder_->ValidateOverflow(size));
+    return Extend(this->value_converter_.get(), obj, size);
+  }
+};
+
+class RStructConverter;
+
+template <>
+struct RConverterTrait<StructType> {
+  using type = RStructConverter;
+};
+
+class RStructConverter : public StructConverter<RConverter, RConverterTrait> {
+ public:
+  Status Append(RScalar* value) override {
+    return Status::NotImplemented("RStructConverter does not use Append()");
+  }
+
+  Status Reserve(int64_t additional_capacity) override {
+    // in contrast with StructConverter, this does not Reserve()
+    // on children, because it will be done as part of Visit() > Extend()
+    return this->builder_->Reserve(additional_capacity);
+  }
+
+  Status Visit(SEXP x, R_xlen_t size) {
+    // iterate over columns of x
+    R_xlen_t n_columns = XLENGTH(x);
+    if (!Rf_inherits(x, "data.frame")) {
+      return Status::Invalid("Can only convert data frames to Struct type");
+    }
+
+    auto struct_builder = checked_cast<StructBuilder*>(this->builder().get());
```

Comment:
You should be able to use `this->struct_builder_`

Reply:
thanks